### DISCUSSION TITLE
Early Stop

### DISCUSSION THREAD

**Discussion by Hua-rookie** [ContinualAI/avalanche#1195] [Q&A] [answered]
I am trying to use the _EarlyStoppingPlugin_ in a ten tasks scenario, however, the 'best_state' and 'best_val' cannot be updated correctly. My code looks like this:
`results = [strategy.eval(test_stream)]`
`for train_task, test_task in zip(train_stream, test_stream):`
      `strategy.train(experiences=train_task, eval_streams=[test_task])`
      `results.append(strategy.eval(test_stream))`

<img width="1079" alt="截屏2022-11-14 21 46 11" src="https://user-images.githubusercontent.com/63031457/201675910-62ecd846-08a2-4cff-9a3a-3ec04478dd6e.png">

After 'results = [strategy.eval(test_stream)]' being executed, the 'best_state' will remain at 0 while the 'best_val' will remain at the value of 'Top1_Acc_Stream/eval_phase/test_stream/Task009' until the end of the program.
As a result,  the training phase will stop as soon as the epoch_num is equal to the patience.

So I wanna to know how to make the _EarlyStoppingPlugin_ working correctly?

**Comment by Hua-rookie** [comment, marked as answer]
I found that there might be a bug in _EarlyStoppingPlugin_.
I have observed that this bug only occurs when a 'return_task_id' is True, not when it is False, and the reasons are as follows:
<img width="666" alt="截屏2022-11-15 11 05 28" src="https://user-images.githubusercontent.com/63031457/201816670-fc27e90a-d750-4a37-adf9-50266dac4a3a.png">
This part of code is in _early_stopping.py_, we can see that only the last metrics would be update to the 'self.best_val'. 
<img width="630" alt="截屏2022-11-15 11 10 34" src="https://user-images.githubusercontent.com/63031457/201817326-08b73118-2405-4d6d-af5e-e79cb23a45e9.png">
This part of code is in _evaluation.py_, we can see that if we choose to 'return_task_id', the last metrics will remain to be the metric of the last task but not the current task it should be.

I hope my understanding is correct.

**Reply by AntonioCarta** [reply]
yes, there is a problem in the current early stopping implementation. The problem is that it only receives the name of the metric. In task-aware scenarios, the metric name depends on the task, so if you give a metric name for a specific task (`.../Task000`) then it's not going to work in the subsequent tasks.
